### PR TITLE
Added Huffman codec to utils

### PR DESCRIPTION
**Pull request by atgoo**
The codec is intended to be used by compression tools and is designed to be compatible with bit_stream utils, but can be built and used independently.

### CODE REVIEW

File: source/util/bit_stream.h [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #ifndef LIBSPIRV_UTIL_BIT_STREAM_H_
 #define LIBSPIRV_UTIL_BIT_STREAM_H_
 
+#include <algorithm>
```
Comment (dneto0) [inline source/util/bit_stream.h]: For std::reverse.

It turns out we also need cassert to make this self contained.  Can be in another CL.

File: source/util/huffman_codec.h [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+  // Prints the Huffman tree in the following format:
+  // w------w------'x'
+  //        w------'y'
+  // Where w stands for the weight of the node.
+  void PrintTree(std::ostream& out) {
+    PrintTreeInternal(out, root_, 0);
+  }
+
+  void PrintTable(std::ostream& out, bool print_weights = true) {
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Method comment.

File: source/util/huffman_codec.h [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+      parent->right = right;
+      queue.push(parent);
+    }
+
+    // Traverse the tree and form encoding table.
+    CreateEncodingTable();
+  }
+
+  // Prints the Huffman tree in the following format:
+  // w------w------'x'
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Good comment.  Examples are golden!

File: source/util/huffman_codec.h [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        if (node->left)
+          queue.emplace(node->left, code + "0");
+
+        if (node->right)
+          queue.emplace(node->right, code + "1");
+      }
+    }
+  }
+
+  const std::unordered_map<Val, std::pair<uint64_t, size_t>>&
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Method comment.

File: source/util/huffman_codec.h [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    auto it = encoding_table_.find(val);
+    if (it == encoding_table_.end())
+      return false;
+    *bits = it->second.first;
+    *num_bits = it->second.second;
+    return true;
+  }
+
+  // Reads bits one-by-one using callback |read_bit| until a match is found.
+  // Matching value is stored in |val|. Returns false if the stream terminates
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Presumably a decoded false maps to 0 and true maps to 1.  You should say so.

File: source/util/huffman_codec.h [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+      }
+    }
+  }
+
+  const std::unordered_map<Val, std::pair<uint64_t, size_t>>&
+      GetEncodingTable() const {
+    return encoding_table_;
+  }
+
+  // Encodes |val| and stores the value in the lower |num_bits| of |bits|.
```
Comment (dneto0) [inline source/util/huffman_codec.h]: "stores the value" ->  "stores the Huffman code for |val|"  ?

File: source/util/huffman_codec.h [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    assert (0);
+    return false;
+  }
+
+ private:
+  // Huffman tree node.
+  struct Node {
+    Val val = Val();
+    uint32_t weight = 0;
+    uint32_t id = 0;
```
Comment (dneto0) [inline source/util/huffman_codec.h]: What is the meaning of "id" ?
Should say something like "Nodes are assigned sequential ids as they are created, starting with 1.  Ids are used to break ordering ties between values with equal weight."

File: source/util/huffman_codec.h [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    Node* left = nullptr;
+    Node* right = nullptr;
+  };
+
+  static bool LeftIsBigger(const Node* left, const Node* right) {
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Method comment.

File: source/util/huffman_codec.h [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+    if (left->weight == right->weight) {
+      assert (left->id != right->id);
+      return left->id > right->id;
+    }
+    return left->weight > right->weight;
+  }
+
+  // Prints subtree (helper function used by PrintTree).
+  static void PrintTreeInternal(std::ostream& out, Node* node, size_t depth) {
```
Comment (dneto0) [inline source/util/huffman_codec.h]: References normally should be const.  Recommend std::ostream* here.
But this is such a common pattern that we've let it slip by before.   Ok to keep if you really want.

File: source/util/huffman_codec.h [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+    if (!node)
+      return;
+
+    if (!node->right && !node->left) {
+      out << node->val << std::endl;
+    } else {
+      if (node->right) {
+        out << std::setfill('-') << std::left << std::setw(7) << node->right->weight;
```
Comment (dneto0) [inline source/util/huffman_codec.h]: You should restore the fill char and width properties of the stream.  It's good stream hygiene.
E.g. see https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/util/hex_float.h#L719

File: source/util/huffman_codec.h [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+        PrintTreeInternal(out, node->right, depth + 1);
+      }
+
+      if (node->left) {
+        out << std::string(depth * 7, ' ');
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Recommend making a local constant to replace the magic "7" values in this function.

File: source/util/huffman_codec.h [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    // and pushing the root of the new tree in the queue.
+    while (true) {
+      // We push a node at the end of each iteration, so the queue is never
+      // supposed to be empty at this point, unless there are no leaves, but
+      // that case was already handled.
+      assert(!queue.empty());
+      Node* right = queue.top();
+      queue.pop();
+
+      // If the queue is empty at this point, then the last node contains is
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Grammar: delete the "contains" on this line.

File: source/util/huffman_codec.h [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+      // the root of the complete Huffman tree.
+      if (queue.empty()) {
+        root_ = right;
+        break;
+      }
+
+      Node* left = queue.top();
+      queue.pop();
+
+      // Combile left and right into a new tree and push it into the queue.
```
Comment (dneto0) [inline source/util/huffman_codec.h]: "Combine"

File: source/util/huffman_codec.h [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+        out << std::setfill('-') << std::left << std::setw(7) << node->left->weight;
+        PrintTreeInternal(out, node->left, depth + 1);
+      }
+    }
+  }
+
+  // Traverses the Huffman tree and saves paths to the leaves as bit
+  // sequences to encoding_table_.
+  void CreateEncodingTable() {
+    std::queue<std::tuple<Node*, uint64_t, size_t>> queue;
```
Comment (dneto0) [inline source/util/huffman_codec.h]: What are the uint64_t and size_t here? 
Probably better to make a struct so you'd have better names?

Comment (dneto0) [inline source/util/huffman_codec.h]: Looks like the uint64_t is the prefix.  Why are 64 bits going to be enough?

Intuition says a worst case code length would be for a degenerately imbalanced tree.  Why does that not happen?

File: source/util/huffman_codec.h [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        auto insertion_result = encoding_table_.emplace(
+            node->val, std::pair<uint64_t, size_t>(bits, depth));
+        assert(insertion_result.second);
+        (void)insertion_result;
+      } else {
+        if (node->left)
+          queue.emplace(node->left, bits, depth + 1);
+
+        if (node->right)
+          queue.emplace(node->right, bits | (1ULL << depth), depth + 1);
```
Comment (dneto0) [inline source/util/huffman_codec.h]: I'd find it slightly easier to read the 1ULL as uint64_t(1)

File: source/util/huffman_codec.h (hunk 6 above)
Comment (dneto0) [inline source/util/huffman_codec.h]: Explain the return value.

File: source/util/huffman_codec.h (hunk 5 above)
Comment (dneto0) [inline source/util/huffman_codec.h]: Have to explain who the callback works.  There are two bools returned: one by return and the other by pointer param.  I shouldn't have to read the body of this method to know how to make a read_bit callback.

File: source/util/huffman_codec.h (hunk 3 above)
Comment (dneto0) [inline source/util/huffman_codec.h]: Should say that the "1" subtree appears above its  "0" sibling.

File: test/huffman_codec.cpp [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+  std::string bits_;
+  size_t pos_ = 0;
+};
+
+TEST(Huffman, PrintTree) {
+  HuffmanCodec<std::string> huffman(GetTestSet());
+  std::stringstream ss;
+  huffman.PrintTree(ss);
+
+  const std::string expected = std::string(R"(
```
Comment (dneto0) [inline test/huffman_codec.cpp]: Elegant test cases.  I'm glad you put this tree output here first!   I'm sure you're glad too.  :-)

File: source/util/huffman_codec.h (hunk 14 above)
Comment (atgoo) [inline source/util/huffman_codec.h]: I thinks in order to fill 64 bits with imbalanced values we would need something like uint128_t for the histogram counts :)

Comment (atgoo) [inline source/util/huffman_codec.h]: Actually no, it just needs to grow faster than fibonacci. But our counts are uint32_t, so we are safe.

Comment (dneto0) [inline source/util/huffman_codec.h]: I figured that would be the essence of the argument.  You need to say that in a comment in the code.

Furthermore, what if I give you a big set of items, each with weight 0.  That's allowed by the signature of the API.  How do you know you won't get a badly balanced tree?

Comment (atgoo) [inline source/util/huffman_codec.h]: I don't think there is such a thing a poorly balanced Huffman tree. If the frequencies are growing at ~2^n then we would have a ladder, but this can't happen. When all nodes have equal weights, then we have a flat table ordered by ids. All codes would have similar length. That's not an edge case, but rather anticlimactic one.

File: source/util/huffman_codec.h [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+      return;
+
+    const size_t kTextFieldWidth = 7;
+
+    if (!node->right && !node->left) {
+      out << node->val << std::endl;
+    } else {
+      if (node->right) {
+        std::stringstream label;
+        label << std::setfill('-') << std::left << std::setw(kTextFieldWidth)
```
Comment (dneto0) [inline source/util/huffman_codec.h]: Unaddressed comment from before: https://github.com/KhronosGroup/SPIRV-Tools/pull/661#discussion_r121531093

File: source/util/huffman_codec.h (hunk 14 above)
Comment (dneto0) [inline source/util/huffman_codec.h]: If all the weights are 0 then everything is ordered by 0 rather than by weights.  Then I think the queue always picks off the nodes with largest IDs to combine them.  So you get a ladder.  ??

This problem goes away if you have as a precondition to the entire table that the weights are assumed to be strictly positive.

Also, you must put the analysis in the *comments* rather than in the review.

File: source/util/huffman_codec.h (hunk 17 above)
Comment (atgoo) [inline source/util/huffman_codec.h]: These options are not global and the stream object is destroyed anyway.

Comment (dneto0) [inline source/util/huffman_codec.h]: It's a mutation of a the out parameter.   So you if you want to keep it this way then your method comment should mention this side effect.

Comment (dneto0) [inline source/util/huffman_codec.h]: Whoops. Misread that.

**Comment by dneto0** [comment]
Remaining changes required:

- HuffmanCodec constructor comment should say that it assumes the histogram does not have entries with zero counts. If you want to be more relaxed then that, then give bounds somehow. But I'd just recommend being strict.
- CreateEncodingTable implementation comment (around the Context struct, probably) should say that because weights are assumed non-zero, then the huffman tree is nearly balanced, and so 64 bits is sufficient as a path label down to any leaf.

**Comment by atgoo** [comment]
> Remaining changes required:

Done.

**Comment by dneto0** [comment]
Squashed and pushed into master as 40a282961
